group-coordinator/src/main/java/org/apache/kafka/coordinator/group/assignor/GeneralUniformAssignmentBuilder.java

```diff
@@ -14,17 +14,891 @@
+                MemberPair existingPair = removeMovementRecordOfPartition(partition);
+                assert existingPair.dstMemberId.equals(oldOwner);
+                if (!existingPair.srcMemberId.equals(newOwner)) {
+                    // The partition is not moving back to its previous consumer.
+                    addPartitionMovementRecord(partition, new MemberPair(existingPair.srcMemberId, newOwner));
+                }
+            } else
+                addPartitionMovementRecord(partition, pair);
+        }
+
+        private TopicIdPartition getTheActualPartitionToBeMoved(TopicIdPartition partition, String oldOwner, String newOwner) {
+            Uuid topic = partition.topicId();
+
+            if (!partitionMovementsByTopic.containsKey(topic))
+                return partition;
+
+            // This partition was previously moved.
+            if (partitionMovementsByPartition.containsKey(partition)) {
+                assert oldOwner.equals(partitionMovementsByPartition.get(partition).dstMemberId);
+                oldOwner = partitionMovementsByPartition.get(partition).srcMemberId;
+            }
+
+            Map<MemberPair, Set<TopicIdPartition>> partitionMovementsForThisTopic = partitionMovementsByTopic.get(topic);
+            MemberPair reversePair = new MemberPair(newOwner, oldOwner);
+            if (!partitionMovementsForThisTopic.containsKey(reversePair))
+                return partition;
+
+            return partitionMovementsForThisTopic.get(reversePair).iterator().next();
+        }
+    }
+
+    /**
+     * Manages assignments to members based on their current assignment size and maximum allowed assignment size.
+     */
+    protected class AssignmentManager {
+        private final Map<String, MemberAssignmentData> membersWithAssignmentSizes = new HashMap<>();
+
+        /**
+         * Represents the assignment metadata for a member.
+         */
+        private class MemberAssignmentData {
+            final String memberId;
+            int currentAssignmentSize = 0;
+            int maxAssignmentSize;
+
+            /**
+             * Constructs a MemberAssignmentData with the given member Id.
+             *
+             * @param memberId The Id of the member.
+             */
+            MemberAssignmentData(String memberId) {
+                this.memberId = memberId;
+            }
+
+            @Override
+            public boolean equals(Object o) {
+                if (this == o) return true;
+                if (o == null || getClass() != o.getClass()) return false;
+                MemberAssignmentData that = (MemberAssignmentData) o;
+                return memberId.equals(that.memberId);
+            }
+
+            @Override
+            public int hashCode() {
+                return Objects.hash(memberId);
+            }
+
+            @Override
+            public String toString() {
+                return "MemberAssignmentData{" +
```

Comment:
nit: We usually use `()` instead of `{}`.

Reply:
ohhh makes sense okay, these were all autogenerated, maybe I'll try to change the settings :/